File: GemBox.Pdf Examples/Print/PrintInWinFormsVb/Form1.vb
Imports System.Drawing.Printing
Imports System.IO
Imports GemBox.Pdf

Partial Public Class Form1
    Inherits Form

    Dim document As PdfDocument

    Public Sub New()
        ComponentInfo.SetLicense("FREE-LIMITED-KEY")
        InitializeComponent()
    End Sub

    Private Sub Form1_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
        If Me.document IsNot Nothing Then
            Me.document.Close()
        End If
    End Sub

    Private Sub LoadFileMenuItem_Click(sender As Object, e As EventArgs) Handles LoadFileMenuItem.Click

        Dim openFileDialog As New OpenFileDialog()
        openFileDialog.Filter =
            "PDF files (*.pdf)|*.pdf"

        If (openFileDialog.ShowDialog() = DialogResult.OK) Then
            If Me.document IsNot Nothing Then
                Me.document.Close()
            End If

            Me.document = PdfDocument.Load(openFileDialog.FileName)
            Me.ShowPrintPreview()
        End If

    End Sub

    Private Sub PrintFileMenuItem_Click(sender As Object, e As EventArgs) Handles PrintFileMenuItem.Click

        If document Is Nothing Then Return

        Dim printDialog As New PrintDialog() With {.AllowSomePages = True}
        If (printDialog.ShowDialog() = DialogResult.OK) Then

            Dim printerSettings As PrinterSettings = printDialog.PrinterSettings
            Dim printOptions As New PrintOptions()

            ' Set PrintOptions properties based on PrinterSettings properties.
            printOptions.CopyCount = printerSettings.Copies
            printOptions.FromPage = printerSettings.FromPage - 1
            printOptions.ToPage = If(printerSettings.ToPage = 0, Integer.MaxValue, printerSettings.ToPage - 1)

            Me.document.Print(printerSettings.PrinterName, printOptions)
        End If

    End Sub

    Private Sub ShowPrintPreview()

        ' Create image for each Word document's page.
        Dim images As Image() = Me.CreatePrintPreviewImages()
        Dim imageIndex As Integer = 0

        ' Draw each page's image on PrintDocument for print preview.
        Dim printDocument = New PrintDocument()
        AddHandler printDocument.PrintPage,
            Sub(sender, e)
                Using image As Image = images(imageIndex)
                    Dim graphics = e.Graphics
                    Dim region = graphics.VisibleClipBounds

                    ' Rotate image if it has landscape orientation.
                    If image.Width > image.Height Then image.RotateFlip(RotateFlipType.Rotate270FlipNone)

                    graphics.DrawImage(image, 0, 0, region.Width, region.Height)
                End Using

                imageIndex += 1
                e.HasMorePages = imageIndex < images.Length
            End Sub

        Me.PageUpDown.Value = 1
        Me.PageUpDown.Maximum = images.Length
        Me.printPreviewControl.Document = printDocument

    End Sub

    Private Function CreatePrintPreviewImages() As Image()

        Dim pageCount As Integer = Me.document.Pages.Count
        Dim images = New Image(pageCount - 1) {}

        For pageIndex As Integer = 0 To pageCount - 1
            Dim imageStream = New MemoryStream()
            Dim imageOptions = New ImageSaveOptions(ImageSaveFormat.Png) With {.PageNumber = pageIndex}

            Me.document.Save(imageStream, imageOptions)
            images(pageIndex) = Image.FromStream(imageStream)
        Next

        Return images

    End Function

    Private Sub PageUpDown_ValueChanged(sender As Object, e As EventArgs) Handles PageUpDown.ValueChanged
        Me.printPreviewControl.StartPage = Me.PageUpDown.Value - 1
    End Sub

End Class

File: C#/Common Uses/Print/Print in WinForms/Form1.cs
using System;
using System.Drawing;
using System.Drawing.Printing;
using System.IO;
using System.Windows.Forms;
using GemBox.Pdf;

public partial class Form1 : Form
{
    PdfDocument document;

    public Form1()
    {
        ComponentInfo.SetLicense("FREE-LIMITED-KEY");
        InitializeComponent();
    }

    void Form1_FormClosing(object sender, FormClosingEventArgs e) => document?.Close();

    void LoadFileMenuItem_Click(object sender, EventArgs e)
    {
        var openFileDialog = new OpenFileDialog
        {
            Filter =
            "PDF files (*.pdf)|*.pdf"
        };

        if (openFileDialog.ShowDialog() == DialogResult.OK)
        {
            document?.Close();

            document = PdfDocument.Load(openFileDialog.FileName);
            ShowPrintPreview();
        }
    }

    void PrintFileMenuItem_Click(object sender, EventArgs e)
    {
        if (document == null)
        {
            return;
        }

        var printDialog = new PrintDialog() { AllowSomePages = true };
        if (printDialog.ShowDialog() == DialogResult.OK)
        {
            PrinterSettings printerSettings = printDialog.PrinterSettings;
            var printOptions = new PrintOptions
            {
                // Set PrintOptions properties based on PrinterSettings properties.
                CopyCount = printerSettings.Copies,
                FromPage = printerSettings.FromPage - 1,
                ToPage = printerSettings.ToPage == 0 ? int.MaxValue : printerSettings.ToPage - 1
            };

            document.Print(printerSettings.PrinterName, printOptions);
        }
    }

    void ShowPrintPreview()
    {
        // Create image for each Word document's page.
        Image[] images = CreatePrintPreviewImages();
        var imageIndex = 0;

        // Draw each page's image on PrintDocument for print preview.
        var printDocument = new PrintDocument();
        printDocument.PrintPage += (sender, e) =>
        {
            using (Image image = images[imageIndex])
            {
                Graphics graphics = e.Graphics;
                RectangleF region = graphics.VisibleClipBounds;

                // Rotate image if it has landscape orientation.
                if (image.Width > image.Height)
                {
                    image.RotateFlip(RotateFlipType.Rotate270FlipNone);
                }

                graphics.DrawImage(image, 0, 0, region.Width, region.Height);
            }

            ++imageIndex;
            e.HasMorePages = imageIndex < images.Length;
        };

        PageUpDown.Value = 1;
        PageUpDown.Maximum = images.Length;
        PrintPreviewControl.Document = printDocument;
    }

    Image[] CreatePrintPreviewImages()
    {
        var pageCount = document.Pages.Count;
        var images = new Image[pageCount];

        for (var pageIndex = 0; pageIndex < pageCount; ++pageIndex)
        {
            var imageStream = new MemoryStream();
            var imageOptions = new ImageSaveOptions(ImageSaveFormat.Png) { PageNumber = pageIndex };

            document.Save(imageStream, imageOptions);
            images[pageIndex] = Image.FromStream(imageStream);
        }

        return images;
    }

    void PageUpDown_ValueChanged(object sender, EventArgs e) => PrintPreviewControl.StartPage = (int)PageUpDown.Value - 1;
}